File: tests/e2e/go_client/row_access_session_user_test.go
// E2E: SESSION_USER() inside a RAP filter predicate (ADR 0038), exercised
// through the official cloud.google.com/go/bigquery Go client.
//
// The canonical "tenant isolation by email domain" pattern:
//   - Seed a tenants table with rows for two domains.
//   - Create a RAP filter
//     REGEXP_EXTRACT(SESSION_USER(), r'@(.+)$') = tenant_id granted to
//     allAuthenticatedUsers.
//   - Each caller sees only their own tenant's rows.
//
// The X-Bqemu-Caller header is injected via the same custom
// http.RoundTripper as row_access_test.go.

package e2e

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"testing"
	"time"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
)

const sessionUserProject = "e2e-go-row_access_session_user"

func sessionUserClient(ctx context.Context, t *testing.T, caller string) *bigquery.Client {
	t.Helper()
	httpClient := &http.Client{
		Transport: &callerHeaderTransport{caller: caller, inner: http.DefaultTransport},
	}
	client, err := bigquery.NewClient(
		ctx,
		sessionUserProject,
		option.WithEndpoint(bqAPIBase()),
		option.WithoutAuthentication(),
		option.WithHTTPClient(httpClient),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return client
}

// Bounded REST client so a stalled emulator can't hang the test run
// indefinitely (CodeRabbit thread PRRT_kwDOSkfuJ86EVwO0). The 15s cap
// matches the per-call SLA the rest of the e2e suite assumes.
var sessionUserHTTP = &http.Client{Timeout: 15 * time.Second}

func sessionUserRest(ctx context.Context, t *testing.T, method, path string, body any) {
	t.Helper()
	var rd io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			t.Fatalf("marshal: %v", err)
		}
		rd = bytes.NewReader(buf)
	}
	req, err := http.NewRequestWithContext(ctx, method, restURL()+path, rd)
	if err != nil {
		t.Fatalf("new request: %v", err)
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := sessionUserHTTP.Do(req)
	if err != nil {
		t.Fatalf("rest %s %s: %v", method, path, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 && resp.StatusCode != http.StatusNoContent {
		b, _ := io.ReadAll(resp.Body)
		t.Fatalf("rest %s %s -> %d: %s", method, path, resp.StatusCode, b)
	}
}

func sessionUserRunSql(ctx context.Context, t *testing.T, client *bigquery.Client, sql string) {
	t.Helper()
	q := client.Query(sql)
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query %q: %v", sql, err)
	}
	var row []bigquery.Value
	for {
		if err := it.Next(&row); err == iterator.Done {
			return
		} else if err != nil {
			t.Fatalf("read %q: %v", sql, err)
		}
	}
}

func sessionUserCollectInts(ctx context.Context, t *testing.T, client *bigquery.Client, sql string) []int64 {
	t.Helper()
	q := client.Query(sql)
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query %q: %v", sql, err)
	}
	var out []int64
	for {
		var row []bigquery.Value
		if err := it.Next(&row); err == iterator.Done {
			return out
		} else if err != nil {
			t.Fatalf("read %q: %v", sql, err)
		}
		out = append(out, row[0].(int64))
	}
}

func TestRowAccessSessionUserFilter(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	dsID := "session_user_go_ds"

	admin := sessionUserClient(ctx, t, "")
	defer admin.Close()
	_ = admin.Dataset(dsID).DeleteWithContents(ctx)

	if err := admin.Dataset(dsID).Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer admin.Dataset(dsID).DeleteWithContents(ctx)

	sessionUserRunSql(ctx, t, admin,
		fmt.Sprintf("CREATE TABLE `%s.%s.tenants` (id INT64, tenant_id STRING)", sessionUserProject, dsID),
	)
	sessionUserRunSql(ctx, t, admin,
		fmt.Sprintf(
			"INSERT INTO `%s.%s.tenants` VALUES (1, 'example.com'), (2, 'example.com'), "+
				"(3, 'other.com'), (4, 'other.com')",
			sessionUserProject, dsID,
		),
	)

	sessionUserRest(ctx, t, "POST",
		fmt.Sprintf("/bigquery/v2/projects/%s/datasets/%s/tables/tenants/rowAccessPolicies",
			sessionUserProject, dsID),
		map[string]any{
			"rowAccessPolicyReference": map[string]string{
				"projectId": sessionUserProject,
				"datasetId": dsID,
				"tableId":   "tenants",
				"policyId":  "tenant_by_session_user",
			},
			"filterPredicate": "REGEXP_EXTRACT(SESSION_USER(), r'@(.+)$') = tenant_id",
			"grantees":        []string{"allAuthenticatedUsers"},
		})

	// @example.com caller sees only example.com tenant rows.
	exampleClient := sessionUserClient(ctx, t, "user:alice@example.com")
	defer exampleClient.Close()
	got := sessionUserCollectInts(ctx, t, exampleClient,
		fmt.Sprintf("SELECT id FROM `%s.%s.tenants` ORDER BY id", sessionUserProject, dsID))
	want := []int64{1, 2}
	if fmt.Sprint(got) != fmt.Sprint(want) {
		t.Fatalf("@example.com caller: got %v want %v", got, want)
	}

	// @other.com caller sees only other.com tenant rows.
	otherClient := sessionUserClient(ctx, t, "user:bob@other.com")
	defer otherClient.Close()
	got = sessionUserCollectInts(ctx, t, otherClient,
		fmt.Sprintf("SELECT id FROM `%s.%s.tenants` ORDER BY id", sessionUserProject, dsID))
	want = []int64{3, 4}
	if fmt.Sprint(got) != fmt.Sprint(want) {
		t.Fatalf("@other.com caller: got %v want %v", got, want)
	}
}

func TestBareSelectSessionUser(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	client := sessionUserClient(ctx, t, "user:claire@example.com")
	defer client.Close()
	q := client.Query("SELECT SESSION_USER() AS who")
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	var row []bigquery.Value
	if err := it.Next(&row); err != nil {
		t.Fatalf("read: %v", err)
	}
	if row[0].(string) != "claire@example.com" {
		t.Fatalf("got %q, want claire@example.com", row[0])
	}
}

// TestBareSelectCurrentUser exercises CURRENT_USER(), documented as a
// co-equal alias for SESSION_USER() in BigQuery's reference (ADR 0040);
// same caller-identity semantics, same pre-translator substitution.
func TestBareSelectCurrentUser(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	client := sessionUserClient(ctx, t, "user:dani@example.com")
	defer client.Close()
	q := client.Query("SELECT CURRENT_USER() AS who")
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	var row []bigquery.Value
	if err := it.Next(&row); err != nil {
		t.Fatalf("read: %v", err)
	}
	if row[0].(string) != "dani@example.com" {
		t.Fatalf("got %q, want dani@example.com", row[0])
	}
}

// TestBareSelectSessionUserSystemVar pins the @@session.user spelling
// (ADR 0040). The system-variable form resolves via the same path as
// the function form — a future SQLGlot AST change for @@session.user
// would surface here as a test failure rather than silently producing
// the ANONYMOUS_CALLER literal.
func TestBareSelectSessionUserSystemVar(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	client := sessionUserClient(ctx, t, "user:eli@example.com")
	defer client.Close()
	q := client.Query("SELECT @@session.user AS who")
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	var row []bigquery.Value
	if err := it.Next(&row); err != nil {
		t.Fatalf("read: %v", err)
	}
	if row[0].(string) != "eli@example.com" {
		t.Fatalf("got %q, want eli@example.com", row[0])
	}
}

File: tests/e2e/java_client/src/test/java/com/example/RowAccessSessionUserTest.java
package com.example;

import com.google.api.gax.rpc.FixedHeaderProvider;
import com.google.cloud.NoCredentials;
import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.BigQueryOptions;
import com.google.cloud.bigquery.DatasetId;
import com.google.cloud.bigquery.DatasetInfo;
import com.google.cloud.bigquery.FieldValueList;
import com.google.cloud.bigquery.QueryJobConfiguration;
import com.google.cloud.bigquery.TableResult;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * E2E: SESSION_USER() inside a RAP filter predicate (ADR 0038), exercised
 * through the official google-cloud-bigquery Java client.
 *
 * The canonical "tenant isolation by email domain" pattern:
 *   - Seed a tenants table with rows for two domains.
 *   - Create a RAP filter REGEXP_EXTRACT(SESSION_USER(), r'@(.+)$') =
 *     tenant_id granted to allAuthenticatedUsers.
 *   - Each caller sees only their own tenant's rows.
 *
 * The X-Bqemu-Caller header is injected per-client via the BigQuery
 * client's FixedHeaderProvider (same pattern as RowAccessTest).
 */
class RowAccessSessionUserTest {
    private static final String REST_URL = System.getenv("BQEMU_REST_URL") != null
            ? System.getenv("BQEMU_REST_URL")
            : "http://localhost:9050";
    private static final String PROJECT = "e2e-java-row_access_session_user";
    private static final String DATASET = "session_user_java_ds";

    private BigQuery admin;

    private BigQuery clientFor(String caller) {
        BigQueryOptions.Builder b = BigQueryOptions.newBuilder()
                .setProjectId(PROJECT)
                .setHost(REST_URL)
                .setCredentials(NoCredentials.getInstance());
        if (caller != null) {
            Map<String, String> headers = new HashMap<>();
            headers.put("X-Bqemu-Caller", caller);
            b.setHeaderProvider(FixedHeaderProvider.create(headers));
        }
        return b.build().getService();
    }

    @BeforeEach
    void setUp() throws Exception {
        admin = clientFor(null);
        try {
            admin.delete(DatasetId.of(PROJECT, DATASET),
                    BigQuery.DatasetDeleteOption.deleteContents());
        } catch (Exception ignore) {
        }
        admin.create(DatasetInfo.newBuilder(DATASET).setLocation("US").build());

        runAdminQuery("CREATE TABLE `" + PROJECT + "." + DATASET + ".tenants` "
                + "(id INT64, tenant_id STRING)");
        runAdminQuery("INSERT INTO `" + PROJECT + "." + DATASET + ".tenants` "
                + "VALUES (1, 'example.com'), (2, 'example.com'), "
                + "(3, 'other.com'), (4, 'other.com')");

        // Defensive: drop a stale RAP before creating, mirroring
        // RowAccessTest.setUp's pattern — DELETE /datasets/X doesn't
        // currently cascade to row access policies.
        restIgnore4xx("DELETE",
                "/bigquery/v2/projects/" + PROJECT + "/datasets/" + DATASET
                        + "/tables/tenants/rowAccessPolicies/tenant_by_session_user",
                null);
        rest("POST",
                "/bigquery/v2/projects/" + PROJECT + "/datasets/" + DATASET
                        + "/tables/tenants/rowAccessPolicies",
                "{\"rowAccessPolicyReference\":{\"projectId\":\"" + PROJECT
                        + "\",\"datasetId\":\"" + DATASET
                        + "\",\"tableId\":\"tenants\",\"policyId\":\"tenant_by_session_user\"},"
                        + "\"filterPredicate\":"
                        + "\"REGEXP_EXTRACT(SESSION_USER(), r'@(.+)$') = tenant_id\","
                        + "\"grantees\":[\"allAuthenticatedUsers\"]}");
    }

    @AfterEach
    void tearDown() {
        try {
            admin.delete(DatasetId.of(PROJECT, DATASET),
                    BigQuery.DatasetDeleteOption.deleteContents());
        } catch (Exception ignore) {
        }
    }

    private TableResult runAdminQuery(String sql) throws Exception {
        return admin.query(QueryJobConfiguration.newBuilder(sql).setUseLegacySql(false).build());
    }

    private TableResult runAs(BigQuery bq, String sql) throws Exception {
        return bq.query(QueryJobConfiguration.newBuilder(sql).setUseLegacySql(false).build());
    }

    // Connect-side timeout on the client, per-request timeout via the
    // HttpRequest builder. The 15s cap matches the per-call SLA the
    // rest of the e2e suite assumes; without it, a stalled emulator
    // can hang the test run indefinitely
    // (CodeRabbit thread PRRT_kwDOSkfuJ86EVwO9).
    private static final Duration REST_TIMEOUT = Duration.ofSeconds(15);
    private static final HttpClient HTTP = HttpClient.newBuilder()
            .version(HttpClient.Version.HTTP_1_1)
            .connectTimeout(REST_TIMEOUT)
            .build();

    private void rest(String method, String path, String body) throws Exception {
        URI uri = URI.create(REST_URL + path);
        HttpRequest.BodyPublisher publisher = body == null
                ? HttpRequest.BodyPublishers.noBody()
                : HttpRequest.BodyPublishers.ofString(body);
        HttpRequest req = HttpRequest.newBuilder(uri)
                .timeout(REST_TIMEOUT)
                .method(method, publisher)
                .header("Content-Type", "application/json")
                .build();
        HttpResponse<String> resp = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
        int rc = resp.statusCode();
        if (rc >= 400 && rc != 204) {
            throw new IOException(
                    "REST " + method + " " + path + " -> " + rc + ": " + resp.body());
        }
    }

    private void restIgnore4xx(String method, String path, String body) throws Exception {
        URI uri = URI.create(REST_URL + path);
        HttpRequest.BodyPublisher publisher = body == null
                ? HttpRequest.BodyPublishers.noBody()
                : HttpRequest.BodyPublishers.ofString(body);
        HttpRequest req = HttpRequest.newBuilder(uri)
                .timeout(REST_TIMEOUT)
                .method(method, publisher)
                .header("Content-Type", "application/json")
                .build();
        HttpResponse<String> resp = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
        int rc = resp.statusCode();
        if (rc >= 500) {
            throw new IOException(
                    "REST " + method + " " + path + " -> " + rc + ": " + resp.body());
        }
    }

    @Test
    void exampleComCallerSeesOnlyExampleComRows() throws Exception {
        BigQuery bq = clientFor("user:alice@example.com");
        TableResult res = runAs(bq,
                "SELECT id FROM `" + PROJECT + "." + DATASET + ".tenants` ORDER BY id");
        List<Long> ids = new ArrayList<>();
        for (FieldValueList row : res.iterateAll()) {
            ids.add(row.get("id").getLongValue());
        }
        assertEquals(java.util.Arrays.asList(1L, 2L), ids);
    }

    @Test
    void otherComCallerSeesOnlyOtherComRows() throws Exception {
        BigQuery bq = clientFor("user:bob@other.com");
        TableResult res = runAs(bq,
                "SELECT id FROM `" + PROJECT + "." + DATASET + ".tenants` ORDER BY id");
        List<Long> ids = new ArrayList<>();
        for (FieldValueList row : res.iterateAll()) {
            ids.add(row.get("id").getLongValue());
        }
        assertEquals(java.util.Arrays.asList(3L, 4L), ids);
    }

    @Test
    void bareSelectSessionUserReturnsBareEmail() throws Exception {
        // Not a RAP test, but uses the same pre-translator substitution
        // path — pinning here keeps both surfaces (RAP filter +
        // free-form query) in one place.
        BigQuery bq = clientFor("user:claire@example.com");
        TableResult res = runAs(bq, "SELECT SESSION_USER() AS who");
        String got = null;
        for (FieldValueList row : res.iterateAll()) {
            got = row.get("who").getStringValue();
            break;
        }
        assertEquals("claire@example.com", got);
    }

    @Test
    void bareSelectCurrentUserReturnsBareEmail() throws Exception {
        // CURRENT_USER() is documented as a co-equal alias for
        // SESSION_USER() in BigQuery's reference (ADR 0040); same
        // caller-identity semantics, same pre-translator substitution.
        BigQuery bq = clientFor("user:dani@example.com");
        TableResult res = runAs(bq, "SELECT CURRENT_USER() AS who");
        String got = null;
        for (FieldValueList row : res.iterateAll()) {
            got = row.get("who").getStringValue();
            break;
        }
        assertEquals("dani@example.com", got);
    }

    @Test
    void bareSelectSessionUserSystemVarReturnsBareEmail() throws Exception {
        // The system-variable spelling @@session.user resolves via the
        // same path as the function form (ADR 0040) — pinned here so a
        // future SQLGlot AST change for @@session.user would surface as
        // a test failure rather than silently producing the
        // ANONYMOUS_CALLER literal.
        BigQuery bq = clientFor("user:eli@example.com");
        TableResult res = runAs(bq, "SELECT @@session.user AS who");
        String got = null;
        for (FieldValueList row : res.iterateAll()) {
            got = row.get("who").getStringValue();
            break;
        }
        assertEquals("eli@example.com", got);
    }
}
